LUA_ERRGCMM, luaL_loadbufferx, luaL_loadfilex

ThePhD opened the pull request:
This pull request implements several changes to address builds on MinGW, Android, and Visual C++ builds trying to avoid deprecation errors and warnings, as well as compiler errors from missing functions. This is the full code necessary to allow sol2 to directly use this compatibility shim, as per @daurnimator's request of me.

It addresses #29 and #30.

Firstly, this implementation defines a few helpers made static in this implementation:
* `COMPAT53_LUA_FILE_BUFFER_SIZE` macro, for buffered reads using the above structure and function
* `compat53_strerror` allows us to normalize differences between implementations and avoid linter error checks
* `COMPAT53_FOPEN_NO_LOCK` macro, see discussion below of `luaL_loadfilex`


# Implementation notes

* For `luaL_fileresult`

Usage of `strerror` directly in `luaL_fileresult` results in a deprecation error. Because this deprecation error's "turn me off" macros and others may change the library's looks and this may be included in header-only projects which do not have the privilege of dictating to the user what they can and cannot define when working with the code, I have added a check for `_MSC_VER` to check for VC++ and thusly call `strerror_s`. This is all put inside of `compat53_strerror`.

* For `luaL_loadfilex`

We take Lua 5.3's implementation, with a few additional checks added since Lua 5.3's `loadfilex` implementation depends on `lua_load` checking the `mode` parameter (and in 5.1 it doesn't take a load parameter), so we do that with our own `compat53_checkmode`.

Additionally, `COMPAT53_FOPEN_NO_LOCK` defined to `0` changes the usage of MSVC's `_fsopen` to `fopen_s`. The reason this is an explicit opt-in macro is because the default behavior of `fopen` (and the reason Microsoft considers it unsafe, along with other reasons) is that `fopen` does not lock the file for reading or writing. `fopen_s` does, but we want to maintain the same semantics as Lua would have if it opened the file: therefore, this macro is an opt-in macro, that defaults to being defined to `1` and using the internal function `_fsopen` with the flag `_SHDENYNO` to have the exact same semantics as `fopen` without the error killing the build.

* For `luaL_loadbufferx`

Similar to  `luaL_loadfilex`, without as much effort for having to open the file to check for the first character. The size is checked and the buffer validated for being a text vs. binary, before continuing on with the processing.

* For `lua_len`

The patch for this function specifically avoids a warning since `lua_pushnumber` takes a `lua_Number` (double) and there is a conversion warning when going from the casted-to `lua_Integer`. I changed the cast to `lua_Number` to avoid the warning in VC++.


I tried to keep the implementation as clean as possible. If you have any questions or comments, please comment here or at the line of code with the error and I will do my best to fix it.

Thank you for your time.

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -14,6 +15,61 @@
+
+#if defined(__GNUC__) && (__GNUC__ < 6)  && defined(__MINGW32__)
+// MinGW 5.x.x and below
+// So as #485 indicates, MinGW doesn't have this function of 5.3 of MinGW
```
grammar

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -336,22 +392,155 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
+  return p->buffer;
+}
+
+COMPAT53_API int luaL_loadfilex(lua_State *L, const char *filename, const char *mode) {
```
I'd prefer that we copied lua 5.3's actual implementation. see http://www.lua.org/source/5.3/lauxlib.c.html#luaL_loadfilex

ThePhD replied:
That's fine, too: I can remove my implementation from the PR. I'm not exactly confident I can copy-paste that exactly as I'm not familiar with the Lua sources.

daurnimator commented on c-api/compat-5.3.h:
```diff
@@ -88,6 +86,13 @@ extern "C" {
 #  define LUA_OPLE 2
 #endif
 
+// LuaJIT/Lua 5.1 does not have the updated 
```
Use `/*` style comments for consistency and compatibility.

ThePhD replied:
I went through and fixed all the comments.

daurnimator commented on c-api/compat-5.3.h:
```diff
@@ -88,6 +86,13 @@ extern "C" {
+// error codes for thread status/function returns (but some patched versions do)
+// define it only if it's not found
+#if !defined(LUA_ERRGCMM)
+#  define LUA_ERRGCMM (LUA_ERRERR + 2) // + 2 because in some version something is already defined at LUA_ERRERR + 1
```
> some version something

Which version, what thing?

ThePhD replied:
I clarified the comment.

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -14,6 +15,61 @@
+
+#if defined(__GNUC__) && (__GNUC__ < 6)  && defined(__MINGW32__)
+// MinGW 5.x.x and below
+// So as #485 indicates, MinGW doesn't have this function of 5.3 of MinGW
```
#485 of what?

mrgreywater replied:
probably https://github.com/ThePhD/sol2/issues/485, should be added to the source comment

ThePhD replied:
I clarified the source.

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -14,6 +15,61 @@
+	// pthreads isn't guaranteed either
+	// Now. The stdc lib devs probably use thread_local on their stuff
+	// but just in case, we'll do this here to be super safe
+	// this should protect us from data in-fighting across threads
```
This doesn't seem correct. `__thread` only makes sense for static variables.
In other cases the variable is on the stack, and hence is already thread-safe.

Also note that this only makes the *pointer* safe, not what it points to.

mrgreywater replied:
`thread_local` and `static thread_local` does the technically the same

ThePhD replied:
Then there's very little I can do. Since I can't get `sys_errlist` to show up on MinGW versions either this is probably the best I can do, unless there's a specific header or something related to the std lib I can find to include that will give me access to it.

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -336,22 +392,155 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
+    char buf[512];
+    strerror_r(en, buf, 1024);
+    s = buf;
+#elif defined(_MSC_VER)
```
This should be a check for `_WIN32`, not the compiler in use.

mrgreywater replied:
@daurnimator Are you sure all compilers and versions on windows support strerror_s? msvc and the latest mingw does, but on mingw you would probably use strerror_r.

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -336,22 +392,155 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
+#if defined(__GLIBC__) || defined(_POSIX_VERSION) || defined(__APPLE__) || (defined(__GNUC__) && (__GNUC__ < 6) && defined (__MINGW32__))
+    // use strerror_r here, because it's available on these specific platforms
+    char buf[512];
+    strerror_r(en, buf, 1024);
```
You're passing 1024 as the length, but the buffer is only 512 long.

mrgreywater replied:
also use the return value, see http://www.club.cc.cmu.edu/~cmccabe/blog_strerror.html

mrgreywater replied:
Also, why doesn't use strerror_r always, you're defining it above for the platforms that don't have it already, right?

ThePhD replied:
`sys_errlist` isn't available everywhere either (I ran into compiler troubles trying to implement it based on that before, too).

mrgreywater replied:
Well, then you could initialize your buffer with `snprintf(str, str_len, "Unknown error %d", err);`, or at least null-terminate it before your call to strerror_r, if you don't want to handle it differently depending on the GNU or posix implementation.

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -336,22 +392,155 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
+COMPAT53_API int luaL_loadbufferx(lua_State *L, const char *buff, size_t sz, const char *name, const char *mode) {
+  size_t modesize;
+  FILE* fp;
+  errno_t err
```
`errno_t` isn't valid in C.

mrgreywater replied:
it is in c11, should still be `int` for c99 and prior.

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -336,22 +392,155 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
+  size_t modesize;
+  FILE* fp;
+  errno_t err
+  int allowbinary = 0;
```
This is all very verbose. Use something more like the check in lua itself: https://www.lua.org/source/5.3/ldo.c.html#checkmode
e.g.
```C
static void checkmode (lua_State *L, const char *mode, const char *x) {
    if (mode && strchr(mode, x[0]) == NULL) {
        lua_pushfstring(L, "attempt to load a %s chunk (mode is '%s')", x, mode);
        return LUA_ERRSYNTAX;
    }
    return LUA_OK;
}

COMPAT53_API int luaL_loadbufferx(lua_State *L, const char *buff, size_t sz, const char *name, const char *mode) {
    int status = LUA_OK;
    if (buff[0] == '\x1b') {
        status = checkmode(L, mode, "binary");
    } else {
        status = checkmode(L, mode, "text");
    }
    if (status != LUA_OK)
        return status;
    return luaL_loadbuffer(L, buff, sz, name);
}
```

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -336,22 +392,155 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
+
+COMPAT53_API int luaL_loadbufferx(lua_State *L, const char *buff, size_t sz, const char *name, const char *mode) {
+  size_t modesize;
+  FILE* fp;
```
`fp` is unused.

mrgreywater commented on c-api/compat-5.3.c:
```diff
@@ -336,22 +392,157 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
+    strerror_s(buf, 1024, en);
+    s = buf;
+#else
+    s = strerror(en);
```
maybe add at least a comment that this is unsafe since it's not thread-safe, or implement your own version (see comment above)

mrgreywater commented on c-api/compat-5.3.c:
```diff
@@ -336,22 +392,157 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
+#elif defined(_MSC_VER)
+    // for MSVC and others, use strerror_s since it's provided by default by the libraries
+    char buf[512];
+    strerror_s(buf, 1024, en);
```
another size mismatch

mrgreywater commented on c-api/compat-5.3.c:
```diff
@@ -14,6 +15,61 @@
+#else
+	__thread 
+#endif // VC++ vs.everyone else
+	char* hopeitsthreadlocal = strerror(errnum);
```
strerror is not threadsafe, maybe use [sys_errlist](http://www.club.cc.cmu.edu/~cmccabe/blog_strerror.html) or strerror_s/strerror_r, as below in the PR.

mrgreywater commented on c-api/compat-5.3.c:
```diff
@@ -401,16 +392,22 @@ COMPAT53_API int luaL_fileresult (lua_State *L, int stat, const char *fname) {
+    /* use strerror_r here, because it's available on these specific platforms */
     char buf[512];
-    strerror_r(en, buf, 1024);
+    strerror_r(en, buf, 512);
```
strerror_r may not modify the buffer on the XSI-compliant libstdc implementation, so buf needs to be properly initialized.

mrgreywater commented on c-api/compat-5.3.c:
```diff
@@ -336,23 +383,124 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
   }
   else {
     lua_pushnil(L);
+#if defined(__GLIBC__) || defined(_POSIX_VERSION) || defined(__APPLE__) || (defined(__GNUC__) && (__GNUC__ < 6) && defined (__MINGW32__))
```
`(defined(__GNUC__) && (__GNUC__ < 6) && defined (__MINGW32__)` shouldn't `strerror_r` also be used when `__GNUC__ >= 6` ?

ThePhD replied:
My patch is only for MinGW 5.x. I have not thoroughly tested 6 or 7, but it is said to have `strerror_r` in the implementation because this code never errored on my 6 or 7 MinGW versions when I ran the tests otherwise. If someone can show me MinGW 6 or 7 having it (there were patches submitted to MinGW implementing `strerror_r`, but I don't know if they were accepted), I can change the conditional.

mrgreywater replied:
Your 'patch' uses strerror, so why don't you just remove '(defined(__GNUC__) && (__GNUC__ < 6) && defined (__MINGW32__)' altogether, and let it fallthrough to strerror here?

ThePhD replied:
Because the maybe-unsafe version is a last resort, to avoid tipping off linters, hyper-pedantic warning/error checkers, and others.

mrgreywater replied:
On mingw it uses strerror either way though right now, so your custom implementation doesn't make any difference.

mrgreywater commented on c-api/compat-5.3.c:
```diff
@@ -336,23 +348,132 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
     lua_pushnil(L);
+#if defined(__GLIBC__) || defined(_POSIX_VERSION) || defined(__APPLE__) || (!defined (__MINGW32__) && defined(__GNUC__) && (__GNUC__ < 6))
+    /* use strerror_r here, because it's available on these specific platforms */
+    char buf[512] = {};
```
should be ` = {0};` in c

ThePhD commented:
I think someone else may have to refactor the `luaL_loadfilex` function to be similar to the Lua 5.3 sources. I myself am not exactly sure I want to copy-paste the code from Lua's sources into here; I feel like someone more familiar would have to do that bit for me to make sure I don't ruin anything.

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -336,23 +348,132 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
+#else
+    /* fallback, but
+     * strerror is not guaranteed to be threadsafe due to modifying
+     * errno itself and some impls not locking a static buffer for it
```
`errno` is threadsafe on all known systems.

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -336,23 +348,132 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
+    strerror_r(en, buf, 512);
+    s = buf;
+#else
+	/* GNU-specific which returns const char* */
```
Indentation (uses tab instead of spaces?)

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -336,23 +348,132 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
+#elif defined(_MSC_VER) || (defined(__STDC_VERSION__) && __STDC_VERSION__ >= 201112L)
+    /* for MSVC and other C11 implementations, use strerror_s 
+     * since it's provided by default by the libraries 
+	*/
```
Indentation (uses tab instead of spaces?)

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -336,23 +348,132 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
     lua_pushnil(L);
+#if defined(__GLIBC__) || defined(_POSIX_VERSION) || defined(__APPLE__) || (!defined (__MINGW32__) && defined(__GNUC__) && (__GNUC__ < 6))
+    /* use strerror_r here, because it's available on these specific platforms */
+    char buf[512] = {0};
```
New vars need to be declared at start of a block in C89

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -336,23 +348,132 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
+	/* GNU-specific which returns const char* */
+    s = strerror_r(en, buf, 512);
+#endif
+#elif defined(_MSC_VER) || (defined(__STDC_VERSION__) && __STDC_VERSION__ >= 201112L)
```
`strerror_s` is not a compulsory part of the standard (and infact, glibc doesn't provide it). Looking at  https://stackoverflow.com/a/17960590/282536 is sounds like an additional check of `__STDC_LIB_EXT1__` would be required.

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -336,23 +348,132 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
+#if defined(__GLIBC__) || defined(_POSIX_VERSION) || defined(__APPLE__) || (!defined (__MINGW32__) && defined(__GNUC__) && (__GNUC__ < 6))
+    /* use strerror_r here, because it's available on these specific platforms */
+    char buf[512] = {0};
+#if ((defined(_POSIX_C_SOURCE) && _POSIX_C_SOURCE >= 200112L) || (defined(_XOPEN_SOURCE) || _XOPEN_SOURCE >= 600)) && (!defined(_GNU_SOURCE) || !_GNU_SOURCE)
```
I usually prefer to factor out this sort of thing to a #define like `HAVE_STRERROR_R`. Not sure what lua-compat-5.3 policy for this sort of thing is.

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -336,23 +348,132 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
+    s = buf;
+#else
+    /* fallback, but
+     * strerror is not guaranteed to be threadsafe due to modifying
```
As far as I know, `strerror` is only not thread-safe due to *locales*. ==> If one thread changes the locale to french, then all theads have their error messages changed: this can happen in the middle of a `strerror` call.

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -15,6 +15,25 @@
+#define COMPAT53_HAVE_STRERROR_R_GNU
+#endif // GNU variant strerror_r
+#endif // XSI/Posix vs. GNU strerror_r
+#elif defined(_MSC_VER) || (defined(__STDC_VERSION__) && __STDC_VERSION__ >= 201112L) || defined(__STDC_LIB_EXT1__)
```
Why `elif`?

ThePhD replied:
It either has `strerror_r` or has `strerror_s`: we don't need to know about both.

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -15,6 +15,25 @@
 /* definitions for Lua 5.1 only */
 #if defined(LUA_VERSION_NUM) && LUA_VERSION_NUM == 501
 
+#if defined(__GLIBC__) || defined(_POSIX_VERSION) || defined(__APPLE__) || (!defined (__MINGW32__) && defined(__GNUC__) && (__GNUC__ < 6))
```
Put the #if inside of the `#ifndef`

ThePhD replied:
I'm confused: which part of the `#if` goes inside the `#ifndef` ?

daurnimator replied:
You want users to be able to pass `-DHAVE_STRERROR_R=1` to their compiler.
So you end up with:
```C
#ifndef COMPAT53_HAVE_STRERROR_R
#if defined(__GLIBC__) || defined(_POSIX_VERSION) || defined(__APPLE__) || (!defined (__MINGW32__) && defined(__GNUC__) && (__GNUC__ < 6))
#define COMPAT53_HAVE_STRERROR_R 1
#else
#define COMPAT53_HAVE_STRERROR_R 0
#endif
#endif
```

Which is verbose due to the new GCC7 `-Wexpansion-to-defined` warning.

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -15,6 +15,25 @@
 
+#if defined(__GLIBC__) || defined(_POSIX_VERSION) || defined(__APPLE__) || (!defined (__MINGW32__) && defined(__GNUC__) && (__GNUC__ < 6))
+#ifndef COMPAT53_HAVE_STRERROR_R
+#define COMPAT53_HAVE_STRERROR_R
```
Define to 1

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -350,36 +369,37 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
   else {
     lua_pushnil(L);
-#if defined(__GLIBC__) || defined(_POSIX_VERSION) || defined(__APPLE__) || (!defined (__MINGW32__) && defined(__GNUC__) && (__GNUC__ < 6))
+#if defined(COMPAT53_HAVE_STRERROR_R)
```
Drop the `defined`

ThePhD replied:
`#if COMPAT53_HAVE_STRERROR_R` errors if the `COMPAT53_HAVE_STRERROR_R` is undefined: I don't define it when it doesn't exist. Should I define it to 0 in the `#else` branch?

daurnimator replied:
Generally I do it like this:
```C
#ifndef HAVE_FOO
#define HAVE_FOO (_WIN32)
#endif
```

However there's a new warning in gcc 7 (`-Wexpansion-to-defined`) which warns on the above style if you use e.g. `defined(__GLIBC__)` where I used `_WIN32`, so yes: in some cases you define to `0`.

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -393,7 +413,7 @@ COMPAT53_API int luaL_fileresult (lua_State *L, int stat, const char *fname) {
 
 
-static const char* compat53_f_parser_handler (lua_State *L, void *data, size_t *size) {
+static const char* compat53_f_parser_handler (lua_State *, void *data, size_t *size) {
```
Why this change?

ThePhD replied:
clang and g++ on -Wall -Wpedantic and all of those other settings warns about unused parameters.

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -371,14 +376,14 @@ COMPAT53_API int luaL_fileresult (lua_State *L, int stat, const char *fname) {
   else {
     lua_pushnil(L);
-#if defined(COMPAT53_HAVE_STRERROR_R)
+#if (defined(COMPAT53_HAVE_STRERROR_R) && COMPAT53_HAVE_STRERROR_R)
```
You shouldn't need a `defined` check here.

ThePhD replied:
You're right: I don't need it anymore since I forcefully define `COMPAT53_HAVE_STRERROR_R` to be 0. ... But it's no harm, to leave it, right?

daurnimator replied:
It's cleaner to take it out. Why would you leave do-nothing conditions lying around in your code?

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -371,14 +376,14 @@ COMPAT53_API int luaL_fileresult (lua_State *L, int stat, const char *fname) {
-#if defined(COMPAT53_HAVE_STRERROR_R)
+#if (defined(COMPAT53_HAVE_STRERROR_R) && COMPAT53_HAVE_STRERROR_R)
     /* use strerror_r here, because it's available on these specific platforms */
 #if defined(COMPAT53_HAVE_STRERROR_R_XSI)
```
Get rid of the `defined` here too.

siffiejoe commented on c-api/compat-5.3.c:
```diff
@@ -336,23 +372,134 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
+    /* use strerror_r here, because it's available on these specific platforms */
+#if defined(COMPAT53_HAVE_STRERROR_R_XSI)
+    /* XSI Compliant */
+    strerror_r(en, buf, 512);
```
`sizeof(buf)` is probably better than an explicit `512`.

siffiejoe commented on c-api/compat-5.3.c:
```diff
@@ -336,23 +372,134 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
+static const char* compat53_f_parser_handler (lua_State *L, void *data, size_t *size) {
+  (void)L; /* better fix for unused parameter warnings */
+  struct compat53_f_parser_buffer *p = (struct compat53_f_parser_buffer*)data;
+  size_t readcount = fread(p->buffer + p->start, sizeof(*p->buffer), sizeof(p->buffer), p->file);
```
There's probably a `- p->start` missing somewhere ... (or a `+ p->start` too many).

daurnimator replied:
Lost in comments on this PR is a TODO of moving to lua 5.3's actual `luaL_loadfilex` implementation.

siffiejoe commented on c-api/compat-5.3.c:
```diff
@@ -336,23 +372,134 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
+  (void)L; /* better fix for unused parameter warnings */
+  struct compat53_f_parser_buffer *p = (struct compat53_f_parser_buffer*)data;
+  size_t readcount = fread(p->buffer + p->start, sizeof(*p->buffer), sizeof(p->buffer), p->file);
+  if (ferror(p->file) != 0 || feof(p->file) != 0) {
```
We should only check for error and/or eof, if we can't read a single byte. Otherwise we return what we got.

ThePhD replied:
I restructured the code to read 1 byte (if possible), check for errors, proceed if there are none. We then check for `feof` at the beginning now of this handler function. If reading 1 byte returned EOF, we exit immediately (`feof` tells us to). Otherwise, we read and then check for errors (which does not include EOF): if were successful, we pass that back to Lua to handle. It then continues this pattern and appropriately consumes the whole file (I went through and checked a reading of a few files, just to make sure, and it works appropriately and sol2's tests pass).

siffiejoe commented on c-api/compat-5.3.c:
```diff
@@ -336,23 +372,134 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
+  }
+  struct compat53_f_parser_buffer fbuf;
+  fbuf.file = fp;
+  fbuf.start = 1;
```
Shouldn't we start at `0`?

ThePhD commented:
I think I've covered all of the feedback, so now might be the time for me to properly split the commits for PR as @daurnimator described. The patches now work on all the systems I'm testing on travis, plus the windows tests and builds on MinGW I do (the build was canceled because OSX tests take forever, but every other test passing before the last one is pretty much a surefire sign everything works out well enough): https://travis-ci.org/ThePhD/sol2/builds/273882419

It'll be a bit while I cut the commits into something much more properly managed. Thank you to everyone who reviewed my code and pointed out all of the errors I made.

siffiejoe commented on c-api/compat-5.3.c:
```diff
@@ -336,23 +372,139 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
+
+COMPAT53_API int luaL_loadbufferx (lua_State *L, const char *buff, size_t sz, const char *name, const char *mode) {
+  int status = LUA_OK;
+  if (buff[0] == '\x1b') {
```
Handle zero-length buffer (without a terminating '\0').

daurnimator commented:
> and that should be almost everyhting, except taking the impl of `luaL_loadfilex` directly from Lua itself... which I'm still hesitant to do, because I'm not sure if I can copy that impl correctly or what the repo master has in mind for owning the top-level of code

lua-compat-5.3 already takes a few things directly from lua itself (see lstrlib.c, ltablib.c, lutf8lib.c. along with some of the implementations in compat-5.3.c).
Yet another function copied would not change things.

ThePhD commented:
I updated this PR's info at the top and addressed all the concerns, including using Lua 5.3's implementation of `luaL_loadfilex`, with additional changes to ensure it properly checks `binary` v. `text` mode.

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -336,23 +395,177 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
+  }
+  else {
+    lua_pushfstring(L, "@%s", filename);
+#if defined(_MSC_VER)
```
Should this be conditional on` _MSC_VER`? or `_WIN32`?

`fopen_s` can have different behaviour than `fopen` (e.g. it holds a lock on the file); so I'm not sure we should be using it...

ThePhD replied:
_WIN32 would apply on MinGW, where the deprecation error doesn't exist: the compiler and its associated library is what's providing the error, not Windows.

ThePhD replied:
As for the locking, at that point I'd have to drop down to an even lower API, like `_fsopen` or similar. I don't have any other choice for VC++ users: the code will explicitly error them if they keep using `fopen` as-is, and if I they value the deprecation errors then it will be my library that gets tossed out the window for not playing ball.

daurnimator replied:
> As for the locking, at that point I'd have to drop down to an even lower API, like _fsopen or similar.

That might be better actually, looks like `_fsopen` with `_SH_DENYNO` is the same as `fopen`, but without the warning.

daurnimator replied:
Also, please add a comment here that the workaround is to *avoid a warning* in MSVC version XYZ. (rather than, e.g. working around an bug or platform idiosyncrasy)

ThePhD replied:
It's not just a warning: it's a hard error and a build won't pass in the default VC++ with it, on any warning level.

I added the checks and also default to using `_fsopen` here, but someone can forcefully opt out of it if they expect `fopen_s`'s behavior by default.

daurnimator replied:
> if they expect fopen_s's behavior by default.

Why would they expect this, when lua itself doesn't have this. => we're trying to give lua semantics here, not invent new ones (even if they're better).

ThePhD replied:
That's why it's an explicit opt-in with a macro: if someone wants this behavior, they would need to explicitly ask for it by definition of `COMPAT53_NO_FOPEN_LOCK` to `0`, which will prompt the typical `fopen_s` on their system. Otherwise, it behaves exactly like Lua / like plain `fopen` does.

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -336,23 +395,177 @@ COMPAT53_API void luaL_traceback (lua_State *L, lua_State *L1,
+} compat53_LoadF;
+
+
+static const char *getF (lua_State *L, void *ud, size_t *size) {
```
Missing `compat53_` prefix on the function name

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -14,6 +15,64 @@
+#    define COMPAT53_HAVE_STRERROR_R 1
+#    if ((defined(_POSIX_C_SOURCE) && _POSIX_C_SOURCE >= 200112L) || (defined(_XOPEN_SOURCE) || _XOPEN_SOURCE >= 600)) && (!defined(_GNU_SOURCE) || !_GNU_SOURCE)
+#    ifndef COMPAT53_HAVE_STRERROR_R_XSI
+#      define COMPAT53_HAVE_STRERROR_R_XSI 1
```
Could you `#define COMPAT53_HAVE_STRERROR_R_GNU 0` in this case? (and similar below).

daurnimator commented on c-api/compat-5.3.c:
```diff
@@ -14,6 +15,88 @@
+
+#if defined(_MSC_VER) && COMPAT53_FOPEN_NO_LOCK
+#include <share.h>
+#endif // VC++ _fsopen for share-allowed file read
```
comment style

daurnimator commented:
Could you do the rebase so that your commits are separate logical units?

ThePhD commented:
I'm going to close this pull request and provide individual pull requests for each of the features.
